src/core/processors/poolprocessor.cpp

```
@@ -0,0 +1,142 @@
+        const auto total =
+            std::accumulate(progress.begin(), progress.end(), 0.0f) / progress.size();
+        progressUpdate = dispatchFront([this, p = processor, total]() {
+            if (auto wrapper = p.lock()) {
+                wrapper->processor.progress(this, total);
+            }
+        });
+    }
+}
+
+PoolProcessor::PoolProcessor(pool::Options options, const std::string& identifier,
+                             const std::string& displayName)
+    : Processor(identifier, displayName)
+    , options_{options}
+    , wrapper_{std::make_shared<pool::detail::Wrapper>(*this)}
+    , delay_{std::chrono::milliseconds(500),
+             [wrapper = std::weak_ptr<pool::detail::Wrapper>(wrapper_)]() {
+                 if (auto w = wrapper.lock()) {
+                     auto& p = w->processor;
+
+                     if (p.queuedDispatch() && !p.states_.empty()) return;
+
+                     if (!p.queue_.empty()) {
+                         p.submit(p.queue_.front());
+                         p.queue_.clear();
+                     }
+                 }
+             }} {}
+
+PoolProcessor::~PoolProcessor() { stopJobs(); }
+
+void PoolProcessor::stopJobs() {
+    for (auto& state : states_) {
+        state->stop = true;
+    }
+}
+
+bool PoolProcessor::hasJobs() { return !states_.empty(); }
+
+void PoolProcessor::submit(Submission& job) {
+    job.setupProgress();
+    states_.push_back(job.state);
+    for (auto& task : job.tasks) {
+        getNetwork()->getApplication()->getThreadPool().enqueueRaw(std::move(task));
+    }
+}
+
+void PoolProcessor::invalidate(InvalidationLevel invalidationLevel, Property* source) {
+    if (delayInvalidation()) {
+        notifyObserversInvalidationBegin(this);
+        PropertyOwner::invalidate(invalidationLevel, source);
+        notifyObserversInvalidationEnd(this);
+    } else {
+        Processor::invalidate(invalidationLevel, source);
+    }
+}
+
+void PoolProcessor::handleError() {
```

Is there any way to get information about the error?

---

just rethrow it

---

Rethrow what? There is no info passed into the function. Seems a bit magic to me :)

---

You can rethrow any caught exception

---

We use that in quite a few places to delegate error handling